### PR TITLE
fix: Add validation of which tools were requested by `get_tools`

### PR DESCRIPTION
**Pull request by tushar-composio**
> [!IMPORTANT]
> Adds validation to ensure only requested tools are executed, updating behavior, API, and function signatures in `toolset.py` and `api.py`.
> 
>   - **Behavior**:
>     - Adds validation in `execute_action` in `toolset.py` to ensure actions are requested before execution.
>     - Updates `validate_tools` in `toolset.py` to store requested actions and apps.
>   - **API Changes**:
>     - Modifies `_validate_tools` in `api.py` to handle empty lists for `apps`, `actions`, and `tags`.
>   - **Function Signatures**:
>     - Changes `get_tools` and `register_tools` signatures in multiple plugin files to use non-optional sequences for `actions`, `apps`, and `tags`.
>   - **Misc**:
>     - Adds `requested_actions` and `requested_apps` attributes to `ComposioToolSet` in `toolset.py`.
>     - Updates tests in `test_toolset.py` to cover new validation logic.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ComposioHQ%2Fcomposio&utm_source=github&utm_medium=referral)<sup> for 38644ef5859e995d4ef57244759528c31e4c98e8. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: python/composio/tools/toolset.py [hunk 1]
```diff
@@ -779,6 +784,18 @@ def execute_action(
...
         """
         action = Action(action)
+        if (
+            self.requested_actions is not None
+            and self.requested_apps is not None
+            and action.slug not in self.requested_actions
+            and action.app not in self.requested_apps
+        ):
+            raise ComposioSDKError(
+                f"Action {action.slug} is being called, but was never requested by the toolset."
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:794]: The error message could be more informative. Consider including:
1. The list of allowed actions/apps that were requested
2. A clearer separation between the error description and the explanation
3. A reference to documentation

Suggested format:
```python
raise ComposioSDKError(
    f"Action '{action.slug}' cannot be executed - not in requested tools.\n"
    f"Allowed actions: {self.requested_actions}\n"
    f"Allowed apps: {self.requested_apps}\n"
    "See: https://docs.composio.dev/validation for more details"
)
```

File: python/composio/tools/toolset.py [hunk 2]
```diff
@@ -41,7 +41,7 @@
     TriggerModel,
     TriggerSubscription,
 )
-from composio.client.enums import TriggerType
+from composio.client.enums import TriggerType, app
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py]: There's an unused import `app` that should be removed. The `App` class is already imported in the new import line.

File: python/composio/tools/toolset.py [hunk 3]
```diff
@@ -310,6 +310,11 @@ def _limit_file_search_response(response: t.Dict) -> t.Dict:
         )
         self.max_retries = max_retries
 
+        # To be populated by get_tools(), from within subclasses like
+        # composio_openai's Toolset.
+        self.requested_actions: t.Optional[t.List[str]] = None
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:315]: Consider adding docstrings for the new instance variables to explain:
- When they get populated
- What they represent
- Whether they should be accessed directly
- Their relationship with the validation system

Example:
```python
# To be populated by get_tools(), from within subclasses
self.requested_actions: t.Optional[t.List[str]] = None
"""List of action slugs that were requested through get_tools().
Used for validating tool execution. Should not be modified directly."""

self.requested_apps: t.Optional[t.List[str]] = None
"""List of app slugs that were requested through get_tools().
Used for validating tool execution. Should not be modified directly."""
```

**Comment by shreysingla11** [comment]
### Code Review Summary

Overall, this PR makes good improvements to the toolset validation system. Here's a breakdown:

#### Positive Aspects ✅
- Adds proper validation to ensure tools being executed were requested
- Makes parameter handling more consistent with empty tuples as defaults
- Implements changes consistently across all plugin implementations
- Improves type safety by removing Optional types where not needed

#### Suggestions for Improvement 🔧
1. Error Message Enhancement:
   - Include list of allowed actions/apps in the error message
   - Add clearer separation between error and explanation
   - Add reference to documentation

2. Documentation:
   - Add detailed docstrings for new instance variables
   - Explain when they get populated and their purpose
   - Document whether they should be accessed directly

3. Code Cleanup:
   - Remove unused `app` import from toolset.py
   - Consider adding helper methods to get allowed actions/apps

#### Code Quality Rating: 8/10 ⭐
The changes are well-structured and improve the system's robustness. The suggestions above would make it even better, but they're not critical for functionality.

File: python/composio/tools/toolset.py [hunk 4]
```diff
@@ -328,6 +328,11 @@ def _limit_file_search_response(response: t.Dict) -> t.Dict:
...
+        self.requested_apps: t.Optional[t.List[str]] = None
```
Comment (tushar-composio) [inline python/composio/tools/toolset.py:334]:
```suggestion
```

**Review by angrybayblade — APPROVED** [review]

**Pull request merged**